keps/sig-node/4210-max-image-gc-age/README.md

```diff
@@ -0,0 +1,678 @@
+by the Kubelet, regardless of disk usage.
+
+## Motivation
+
+<!--
+This section is for explicitly listing the motivation, goals, and non-goals of
+this KEP.  Describe why the change is important and the benefits to users. The
+motivation section can optionally provide links to [experience reports] to
+demonstrate the interest in a KEP within the wider Kubernetes community.
+
+[experience reports]: https://github.com/golang/go/wiki/ExperienceReports
+-->
+Currently, all image garbage collection the Kubelet is triggered by disk usage going over a threshold (ImageGCLowThresholdPercent).
+However, there are cases that additional conditions could be considered useful. One such condition is maximum age of an image.
+If an image is unused for a long time (the exact amount of time will be decided, but on the order of weeks is what comes to mind),
+then it is not likely to be used again.
+
+One such condition that can be imagined are clusters with automatic upgrades. If a cluster goes through an upgrade process,
+it will likely have images cached from the old release (old kube-apiserver/etcd/etc). While these images would eventually get removed
+through the disk usage condition, they will needlessly occupy disk space before that.
+
+### Goals
+
+- Introduce an option to the Kubelet ImageMaximumGCAge
+
+### Non-Goals
+
+- Introduce other conditions for image garbage collection to trigger.
+    - A WG was put together in SIG-Node to collect other GC use cases, and no other conditions were identified.
+
+## Proposal
+
+Add an option to the KubeletConfiguration object that looks like:
+```
+	// imageMaximumGCAge is the maximum age an image can be unused before it is garbage collected.
+	ImageMaximumGCAge metav1.Duration
+```
+
+To begin, this option will be set to the maximum duration (roughly 290 years, or the int64 max), effectively disabling it.
+In the future, a more reasonable default may be chosen.
+
+### User Stories (Optional)
+
+#### Story 1
+
+As a cluster admin, I would like my unused images to be garbage collected in a timely manner, and not occupy disk space forever.
+
+### Notes/Constraints/Caveats (Optional)
+
+<!--
+What are the caveats to the proposal?
+What are some important details that didn't come across above?
+Go in to as much detail as necessary here.
+This might be a good place to talk about core concepts and how they relate.
+-->
+
+### Risks and Mitigations
+
+<!--
+What are the risks of this proposal, and how do we mitigate? Think broadly.
+For example, consider both security and how this will impact the larger
+Kubernetes ecosystem.
+
+How will security be reviewed, and by whom?
+
+How will UX be reviewed, and by whom?
+
+Consider including folks who also work outside the SIG or subproject.
+-->
+- If set incorrectly the ImageMaximumGCAge option could cause unneeded image pulls. For instance, if a Cron job ran
```

Comment:
In the CRI API, we have a concept of "Pinned" which would recommend that this image never be GC.

```
// Basic information about a container image.
message Image {
    ...
    // Recommendation on whether this image should be exempt from garbage collection.
    // It must only be treated as a recommendation -- the client can still request that the image be deleted,
    // and the runtime must oblige.
    bool pinned = 8;
}
```

It may be worth a comment that this KEP would avoid any image that is pinned.

Reply:
@kannon92  https://github.com/containerd/containerd/issues/6930

That pinning support for abrbitrary image is still not implemented but is in discussion. Until then, is there not a possibility of this workflow GC'ing an image that is actually required for the cluster ?

Reply:
Actually, scratch that. With https://github.com/containerd/containerd/pull/7944  getting merged, adding a label to the image saying `io.cri-containerd.pinned=pinned` should indicate that the image is pinned when one invokes the image APIs on the cri